source/extensions/filters/network/mongo_proxy/bson_impl.cc

```diff
@@ -368,18 +368,24 @@ std::string FieldImpl::toString() const {
 }
 
 void DocumentImpl::fromBuffer(Buffer::Instance& data) {
-  uint64_t original_buffer_length = data.length();
-  int32_t message_length = BufferHelper::removeInt32(data);
-  if (static_cast<uint64_t>(message_length) > original_buffer_length) {
+  const ssize_t original_buffer_length = data.length();
+  const int32_t message_length = BufferHelper::removeInt32(data);
+  if (message_length <= 0 || message_length > original_buffer_length) {
     throw EnvoyException("invalid BSON message length");
   }
 
   ENVOY_LOG(trace, "BSON document length: {} data length: {}", message_length,
             original_buffer_length);
 
   while (true) {
-    uint64_t document_bytes_remaining = data.length() - (original_buffer_length - message_length);
+    const ssize_t bytes_remaining_after_message = original_buffer_length - message_length;
+    const ssize_t document_bytes_remaining =
+        static_cast<ssize_t>(data.length()) - bytes_remaining_after_message;
     ENVOY_LOG(trace, "BSON document bytes remaining: {}", document_bytes_remaining);
+    // Although mongo_proxy traffic is trusted, do a minimal check.
+    if (document_bytes_remaining < 0) {
```

Comment:
I wonder if it should be `document_bytes_remaining <= 0`.

Reply:
Nvm... on another inspection of the code, it seems that it should throw an exception when calling `BufferHelper::removeByte`.

Reply:
I think having this be `<= 0` makes sense as the document would still be invalid and this would be more specific than the invalid buffer size throw.. all bson should be null terminated.

Reply:
done